core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) {
+    val maxLambdaVar = None
+  }
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+}
+
+case class NormalizePackageMap(pm: PackageMap.Inferred) {
+  import Normalization._
+  import TypedExpr._
+
+  val normalizePackageMap: PackageMap.Normalized = 
+    PackageMap(
+      pm.toMap.toList.map { case (name, pkg) => {
+        normalizePackage(name, pkg).map((name, _))
+    }}.sequence.run(Map()).value._2.toMap)
+
+  def normalizeExpr(expr: TypedExpr[Declaration], env: Env, p: Package.Inferred): State[
+    Map[(PackageName, Identifier), ResultingRef],
+    TypedExpr[(Declaration, NormalExpressionTag)]] =
+      expr match {
+        case a@Annotation(_, _, _) => normalizeAnotation(a, env, p)
+        case g@Generic(_, _, _) => normalizeGeneric(g, env, p)
+        case v@Var(_, _, _, _) => normalizeVar(v, env, p)
+        case al@AnnotatedLambda(_, _, _, _) => normalizeAnnotatedLambda(al, env, p).map(_.asInstanceOf[TypedExpr[(Declaration, NormalExpressionTag)]])
+        case a@App(_, _, _, _) => normalizeApp(a, env, p).map(_.asInstanceOf[TypedExpr[(Declaration, NormalExpressionTag)]])
+        case l@Let(_, _, _, _, _) => normalizeLet(l, env, p)
+        case l@Literal(_, _, _) => normalizeLiteral(l, env, p)
+        case m@Match(_, _, _) => normalizeMatch(m, env, p)
+      }
+
+  private def combineWithChildren(nt: NormalExpressionTag) = nt.children + nt.ne
+
+  def normalizeAnotation(a: Annotation[Declaration], env: Env, p: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  TypedExpr[(Declaration, NormalExpressionTag)]] =
+    for {
+      term <- normalizeExpr(a.term, env, p)
+      neTag = term.tag._2
+      newNeTag = neTag.copy(children = (neTag.children + neTag.ne))
+      tag = (a.tag, newNeTag)
+    }  yield a.copy(term=term, tag=tag)
+
+  def normalizeGeneric(g: Generic[Declaration], env: Env, p: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  TypedExpr[(Declaration, NormalExpressionTag)]] =
+    for {
+      in <- normalizeExpr(g.in, env, p)
+      neTag = in.tag._2
+      newNeTag = neTag.copy(children = neTag.children + neTag.ne)
+      tag = (g.tag, newNeTag)
+    } yield g.copy(in=in, tag=tag)
+
+  def normalizeVar(v: Var[Declaration], env: Env, p: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  TypedExpr[(Declaration, NormalExpressionTag)]] = env._1.get(v.name) match {
+    case None => for {
+      ne <- norm((p, Left((v.name, v.tag)), env))
+      neTag = getTag(ne)._2
+    } yield v.copy(tag=(v.tag, neTag))
+    case Some(neTag) =>
+      State.pure(v.copy(tag=(v.tag, neTag)))
+  }
+
+  def normalizeAnnotatedLambda(al: AnnotatedLambda[Declaration], env: Env, p: Package.Inferred): State[
```

Comment:
I think a `type NormState[A] = State[Map[(PackageName, Identifier), ResultingRef, A]` in the `object Normalization` would help cut down on the type verbosity here.